[SPARK-36834][SHUFFLE] Add support for namespacing log lines emitted by external shuffle service

thejdeep opened the pull request:
### What changes were proposed in this pull request?
Added a config `spark.yarn.shuffle.service.logs.namespace` which can be used to add a namespace suffix to log lines emitted by the External Shuffle Service.

### Why are the changes needed?
Since many instances of ESS can be running on the same NM, it would be easier to distinguish between them.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
N/A

AmplabJenkins commented:
Can one of the admins verify this patch?

thejdeep commented:
cc : @xkrogen @mridulm

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -93,7 +93,7 @@
  */
 public class YarnShuffleService extends AuxiliaryService {
-  private static final Logger logger = LoggerFactory.getLogger(YarnShuffleService.class);
+  private static final Logger logger = null;
```
This should not be `static` or `final` ?

thejdeep replied:
Was a bad merge on my end, thanks. Fixed it.

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -284,7 +293,9 @@ static MergedShuffleFileManager newMergedShuffleFileManagerInstance(TransportCon
-      logger.error("Unable to create an instance of {}", mergeManagerImplClassName);
+      if(logger != null) {
+        logger.error("Unable to create an instance of {}", mergeManagerImplClassName);
+      }
```
When can this be `null` ?

thejdeep replied:
This can be null we use this for running a few tests in the `YarnShuffleServiceSuite`.

mridulm replied:
Oh wait, this is a static method - logger wont be available here.

mridulm commented:
+CC @tgravescs, @Ngone51

tgravescs commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -284,7 +293,6 @@ static MergedShuffleFileManager newMergedShuffleFileManagerInstance(TransportCon
       // will also need the transport configuration.
       return mergeManagerSubClazz.getConstructor(TransportConf.class).newInstance(conf);
     } catch (Exception e) {
-      logger.error("Unable to create an instance of {}", mergeManagerImplClassName);
```
does something else print a useful error message?

thejdeep replied:
Cannot use the logger inside this static context. Therefore, ended up removing it.

mridulm replied:
I agree with @tgravescs, dropping the log message (particularly an error) would be missing out on very useful debugging information.
One option would be to move it out of this class into some other util ?

xkrogen replied:
Perhaps we need to create a static logger (w/o prefix) to fall back to, in the case that we're unable to access the prefixed instance logger?

tgravescs commented:
conceptually looks fine to me too.

xkrogen commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +204,15 @@ protected void serviceInit(Configuration externalConf) throws Exception {
+    String logsNamespace = _conf.get(SPARK_SHUFFLE_SERVICE_LOGS_NAMESPACE_KEY, "");
+    String loggerName = logsNamespace.isEmpty() ? className : className + " " + logsNamespace;
+    logger = LoggerFactory.getLogger(loggerName);
+    logger.info("Initialized Spark YARN shuffle service with configuration overlay from {}",
```
We need a different message if `confOverlayUrl` is null, right? Maybe
```
if (confOverlayUrl == null) {
  logger.info("Initialized Spark YARN shuffle service using base configuration");
} else {
  logger.info(...); <- current logic here
}
```

thejdeep replied:
Added a default message, thanks

xkrogen commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +204,15 @@ protected void serviceInit(Configuration externalConf) throws Exception {
       _conf.addResource(confOverlayUrl);
     }
+    String logsNamespace = _conf.get(SPARK_SHUFFLE_SERVICE_LOGS_NAMESPACE_KEY, "");
+    String loggerName = logsNamespace.isEmpty() ? className : className + " " + logsNamespace;
```
Maybe we can use something slightly more structured than a space

Right now assuming the namespace is "version311" we'll have something like:
```
INFO network.yarn.YarnShuffleService version311 Initialized Spark YARN shuffle ...
```
Maybe this would look better:
```
INFO network.yarn.YarnShuffleService [version311] Initialized Spark YARN shuffle ...
```
Or, since we're considering it part of the name and since frameworks may assume the name looks like a class,
```
INFO network.yarn.YarnShuffleService.version311 Initialized Spark YARN shuffle ...
```
(if we go with the last option because we think following the class-name-like convention is valuable, then we should probably log a warning if the namespace contains spaces)

xkrogen commented on docs/running-on-yarn.md:
```diff
@@ -783,6 +783,14 @@ The following extra configuration options are available when the shuffle service
+  <td><code>spark.yarn.shuffle.service.logs.namespace</code></td>
+  <td><code>(not set)</code></td>
+  <td>
+    A namespace which will be appended to the class name when forming the logger name to use for 
+    emitting logs from the YARN shuffle service.
+  </td>
```
Once we finalize on how we add the namespace to the logger name (per the comment I opened), we should provide an example here of what the final name will look like

thejdeep commented:
Updated the PR with the changes to account for logging in static cases and using period as a delimiter instead of a space. Please review @xkrogen @mridulm Thanks

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +206,24 @@ protected void serviceInit(Configuration externalConf) throws Exception {
-          confOverlayUrl);
       _conf.addResource(confOverlayUrl);
     }
+    String logsNamespace = _conf.get(SPARK_SHUFFLE_SERVICE_LOGS_NAMESPACE_KEY, "");
+    String loggerName = logsNamespace.isEmpty() ? className : className + "." + logsNamespace;
+    logger = LoggerFactory.getLogger(loggerName);
```
Initialize `logger` to `staticLogger` and only create a custom logger in case `logsNamespace != ""` ?

```suggestion
    String logsNamespace = _conf.get(SPARK_SHUFFLE_SERVICE_LOGS_NAMESPACE_KEY, "");
    if (!logsNamespace.isEmpty()) {
      String loggerName = className + "." + logsNamespace;
      logger = LoggerFactory.getLogger(loggerName);
    }
```

thejdeep replied:
Thanks, this is much cleaner.

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -93,7 +93,8 @@
 public class YarnShuffleService extends AuxiliaryService {
-  private static final Logger logger = LoggerFactory.getLogger(YarnShuffleService.class);
+  private static final Logger staticLogger = LoggerFactory.getLogger(YarnShuffleService.class);
+  private Logger logger = null;
```
```suggestion
  private Logger logger = staticLogger;
```

And use logger for all non-static methods.
(See comment below in `serviceInit`).

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -178,7 +185,7 @@ public YarnShuffleService() {
     // until `serviceInit` is called, at which point it's too late to adjust the name.
     super("spark_shuffle");
-    logger.info("Initializing YARN shuffle service for Spark");
+    staticLogger.info("Initializing YARN shuffle service for Spark");
```
We can revert this

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +206,24 @@ protected void serviceInit(Configuration externalConf) throws Exception {
+    String className = YarnShuffleService.class.getName();
     if (confOverlayUrl != null) {
-      logger.info("Initializing Spark YARN shuffle service with configuration overlay from {}",
-          confOverlayUrl);
```
Add this back

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +206,24 @@ protected void serviceInit(Configuration externalConf) throws Exception {
+      logger.warn("Logging namespace contains a space which might " +
+              "not be parseable by some frameworks");
+    }
+    if (confOverlayUrl != null) {
+      logger.info("Initialized Spark YARN shuffle service with configuration overlay from {}",
+              confOverlayUrl);
+    } else {
+      logger.info("Initialized Spark YARN shuffle service with base configuration");
+    }
```
We can remove this, and keep the original code.

```suggestion
```

thejdeep committed:
[SPARK-36834][SHUFFLE] Add support for namespacing log lines emitted by ESS

 ### What changes were proposed in this pull request?
 Added a config `spark.yarn.shuffle.service.logs.namespace` which can be used to add a namespace suffix to log lines emitted by the External Shuffle Service.

 ### Why are the changes needed?
 Since many instances of ESS can be running on the same NM, it would be easier to distinguish between them.

 ### Does this PR introduce _any_ user-facing change?
 No

 ### How was this patch tested?
 N/A

thejdeep commented:
@tgravescs @mridulm @xkrogen Addressed the comments, please review. Thanks

xkrogen commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -199,11 +205,22 @@ protected void serviceInit(Configuration externalConf) throws Exception {
+      logger = LoggerFactory.getLogger(className + "." + logsNamespace);
+    }
+    if (logsNamespace.contains(" ")) {
+      logger.warn("Logging namespace contains a space which might " +
+              "not be parseable by some frameworks");
```
I think we can make the message a bit more actionable by including the config name (`SPARK_SHUFFLE_SERVICE_LOGS_NAMESPACE_KEY`) and explaining the suggested format ("... It is suggested to configure this to a simple string that looks like a Java package or class name to conform to standard logger name format")

xkrogen replied:
Actually while thinking about this further, maybe it doesn't make sense to log a warning. Users may not be leveraging any framework which expects logger names to look like class names, and then they will get an unnecessary warning. Perhaps we can just rely on the note in the config documentation which says to be careful about using spaces. @mridulm what do you think?

xkrogen commented on docs/running-on-yarn.md:
```diff
@@ -793,6 +793,15 @@ The following extra configuration options are available when the shuffle service
+  <td><code>spark.yarn.shuffle.service.logs.namespace</code></td>
+  <td><code>(not set)</code></td>
+  <td>
+    A namespace which will be appended to the class name when forming the logger name to use for
+    emitting logs from the YARN shuffle service. This would throw a warning if the namespace value
+    provided contains a space, for example, "namespace test".
```
Similar to my comment for the warning message, can we do more to explain why we encourage these naming conventions? Maybe:
```
    A namespace which will be appended to the class name when forming the logger name to use for
    emitting logs from the YARN shuffle service, like
    `org.apache.spark.network.yarn.YarnShuffleService.logsNamespaceValue`. Since some logging frameworks
    may expect the logger name to look like a class name, it's generally recommended to provide a value which
    would be a valid Java package or class name and not include spaces.
```

xkrogen reviewed (changes requested):
Logic itself LGTM, I think having the base static logger which we later override is a good approach

I think we can make the documentation/messages a bit more user friendly

mridulm commented on common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:
```diff
@@ -93,7 +93,7 @@
  */
 public class YarnShuffleService extends AuxiliaryService {
-  private static final Logger logger = LoggerFactory.getLogger(YarnShuffleService.class);
+  private static Logger logger = LoggerFactory.getLogger(YarnShuffleService.class);
```
To clarify, this should be `final` - the `defaultLogger` from earlier PR.
Add a new `private Logger logger` - which is initialized to `defaultLogger` by default; and overridden in `serviceInit` as required.

xkrogen reviewed (changes requested):
LGTM except for minor formatting comment

thejdeep commented:
@mridulm @tgravescs Please take a look, thanks!

mridulm reviewed (approved):
+CC @tgravescs, @xkrogen

mridulm commented:
Ok to test

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144271/

SparkQA commented:
**[Test build #144272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144272/testReport)** for PR 34079 at commit [`c3988e9`](https://github.com/apache/spark/commit/c3988e9e7dc68d69ec22dafbcd3499bf3eb80b6e).

SparkQA commented:
**[Test build #144272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144272/testReport)** for PR 34079 at commit [`c3988e9`](https://github.com/apache/spark/commit/c3988e9e7dc68d69ec22dafbcd3499bf3eb80b6e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48751/
